File: Java/380.java
__________________________________________________________________________________________________
sample 50 ms submission
class RandomizedSet {
    private HashMap<Integer, Integer> map;
    private ArrayList<Integer> list;
    java.util.Random rand = new java.util.Random();
    /** Initialize your data structure here. */
    public RandomizedSet() {
        map = new HashMap<>();
        list = new ArrayList<>();
    }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    public boolean insert(int val) {
        if(map.containsKey(val)) return false;
        else{
            map.put(val, list.size());
            list.add(val);
        }
        return true;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    public boolean remove(int val) {
        if(!map.containsKey(val)) return false;
        else{
            int index = map.get(val);
            if(index < list.size() - 1){
                int last = list.get(list.size() - 1);
                list.set(index,last);
                map.put(last,index);
            }
            map.remove(val);
            list.remove(list.size() - 1);
        }
        return true;
    }
    
    /** Get a random element from the set. */
    public int getRandom() {
        return list.get(rand.nextInt(list.size()));
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */
__________________________________________________________________________________________________
sample 40276 kb submission
class RandomizedSet {
    
    Map<Integer, Integer> map;
    List<Integer> list;

    /** Initialize your data structure here. */
    public RandomizedSet() {
        map = new HashMap<>();
        list = new ArrayList<>();
    }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    public boolean insert(int val) {
        boolean isPresent = map.containsKey(val);
        if(!isPresent) {
            list.add(val);
            map.put(val, list.size()-1);
        }
        
        return !isPresent;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    public boolean remove(int val) {
        boolean isPresent = map.containsKey(val);
        if(isPresent) {
            int index = map.get(val);
            if(index == list.size()-1) {
                list.remove(index);
            }else {
                // get the value of swapped item so that we can update its index in the map
                // so that later on if this item gets deleted , index will be correct.
                int swapItem = list.get(list.size()-1);
                Collections.swap(list, index, list.size()-1);
                list.remove(list.size()-1);
                map.put(swapItem, index);
            }
            
            map.remove(val);
        }
        
        return isPresent;
    }
    
    /** Get a random element from the set. */
    public int getRandom() {
       int max = list.size();
        int min = 0;
        int ind = (int)(Math.random() * (max - min) + min);
        return list.get(ind);
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */
__________________________________________________________________________________________________

File: Python3/380.py
__________________________________________________________________________________________________
sample 96 ms submission
import random
class RandomizedSet:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.arr = []
        self.d = dict()
        
        

    def insert(self, val: int) -> bool:
        """
        Inserts a value to the set. Returns true if the set did not already contain the specified element.
        """
        if val in self.d:
            return False
        self.d[val] = len(self.arr)
        self.arr.append(val)     
        return True

    def remove(self, val: int) -> bool:
        """
        Removes a value from the set. Returns true if the set contained the specified element.
        """
        
        if val not in self.d:
            return False
        
        index = self.d[val]
        
        #swap
        temp = self.arr[-1]
        self.arr[-1] = self.arr[index]
        self.arr[index] = temp        
        
        self.d[temp] = index
        
        self.arr.pop()
        del self.d[val]
        return True
    
    def getRandom(self) -> int:
        """
        Get a random element from the set.
        """
        return random.choice(self.arr)
# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
__________________________________________________________________________________________________
sample 16688 kb submission
class RandomizedSet:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.size = 0
        self.array = []
        self.val2i = {}

    def insert(self, val: int) -> bool:
        """
        Inserts a value to the set. Returns true if the set did not already contain the specified element.
        """
        if val in self.val2i: return False
        if self.size == len(self.array): self.array.append(val)
        else: self.array[self.size] = val
        self.val2i[val] = self.size
        self.size += 1
        #print(self.size)
        return True
        
        

    def remove(self, val: int) -> bool:
        """
        Removes a value from the set. Returns true if the set contained the specified element.
        """
        if val not in self.val2i: return False
        #print(self.val2i)
        i = self.val2i[val]
        del self.val2i[val]
        #print(self.val2i)
        if i != self.size - 1:
            self.array[i], self.array[self.size - 1] = self.array[self.size - 1], self.array[i]
            self.val2i[self.array[i]] = i
        self.size -= 1
        
        #print(self.size)
        return True

    def getRandom(self) -> int:
        """
        Get a random element from the set.
        """
        #print(self.size)
        i = random.randrange(self.size)
        return self.array[i]
# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
__________________________________________________________________________________________________
